File: clojure_service/src/cad_os/models/washer.clj
(ns cad-os.models.washer
  (:require [cad-os.commands :as commands]
            [cad-os.models.registry :as registry]
            [cad-os.utils.logger :as logger]))

;; Initialize logger
(def log (logger/get-logger))

;; Schema definition for washer
(def washer-schema
  {:name "Washer"
   :description "A simple washer (ring) with inner and outer diameters"
   :parameters
   [{:name "outer-diameter"
     :type "number"
     :description "Outer diameter of the washer"
     :default 10.0}
    {:name "inner-diameter"
     :type "number"
     :description "Inner diameter of the washer"
     :default 6.0}
    {:name "thickness"
     :type "number"
     :description "Thickness of the washer"
     :default 2.0}
    ;; Optional position parameters (hidden from UI by default)
    {:name "position-x"
     :type "number"
     :description "X position of the washer"
     :default 0.0
     :hidden true}
    {:name "position-y"
     :type "number"
     :description "Y position of the washer"
     :default 0.0
     :hidden true}
    {:name "position-z"
     :type "number"
     :description "Z position of the washer"
     :default 0.0
     :hidden true}]

   ;; Validation rules using expressions
   :validation-rules
   [{:expr "outer-diameter > 0.1"
     :message "Outer diameter must be greater than 0.1"}
    {:expr "inner-diameter > 0.1"
     :message "Inner diameter must be greater than 0.1"}
    {:expr "thickness > 0.1"
     :message "Thickness must be greater than 0.1"}
    {:expr "inner-diameter < outer-diameter"
     :message "Inner diameter must be less than outer diameter"}]})

;; Command generator function for washer
(defn generate-washer-commands
  "Generate commands to create a washer model"
  [params]
  (let [outer-diameter (get params :outer-diameter)
        inner-diameter (get params :inner-diameter)
        thickness (get params :thickness)
        position-x (get params :position-x 0)
        position-y (get params :position-y 0)
        position-z (get params :position-z 0)]

    ((:info log) "Generating washer commands"
                 {:outer outer-diameter
                  :inner inner-diameter
                  :thickness thickness
                  :position [position-x position-y position-z]})

    [(commands/insert-right-circular-cylinder
      "outer" [position-x position-y position-z] [0 0 thickness] (/ outer-diameter 2))
     (commands/insert-right-circular-cylinder
      "inner" [position-x position-y position-z] [0 0 thickness] (/ inner-diameter 2))
     (commands/subtraction "washer" "outer" "inner")]))

;; Register the washer model
(registry/register-model
 "washer"
 washer-schema
 generate-washer-commands)

File: discord-bot/commands/washer.js
const { SlashCommandBuilder, AttachmentBuilder } = require('discord.js');
const cadService = require('../services/cad-service');
const logger = require('../utils/logger')('washer-command');
const { handleCommandError } = require('../utils/errorHandler');

module.exports = {
  data: new SlashCommandBuilder()
    .setName('washer')
    .setDescription('Generate a washer model with specified parameters')
    .addNumberOption(option => 
      option.setName('outer_diameter')
        .setDescription('Outer diameter of the washer')
        .setRequired(true)
        .setMinValue(0.1))
    .addNumberOption(option => 
      option.setName('inner_diameter')
        .setDescription('Inner diameter of the washer')
        .setRequired(true)
        .setMinValue(0.1))
    .addNumberOption(option => 
      option.setName('thickness')
        .setDescription('Thickness of the washer')
        .setRequired(true)
        .setMinValue(0.1)),
  
  async execute(interaction) {
    try {
      logger.info('Washer command execution started', {
        userId: interaction.user.id,
        guildId: interaction.guildId
      });
      
      // Immediately defer the reply to prevent timeout
      await interaction.deferReply();
      logger.debug('Reply deferred');
      
      const outerDiameter = interaction.options.getNumber('outer_diameter');
      const innerDiameter = interaction.options.getNumber('inner_diameter');
      const thickness = interaction.options.getNumber('thickness');
      
      logger.info('Washer parameters received', { 
        outerDiameter, 
        innerDiameter, 
        thickness 
      });
      
      // Validate parameters
      if (innerDiameter >= outerDiameter) {
        logger.warn('Invalid parameters: inner diameter >= outer diameter', {
          innerDiameter,
          outerDiameter
        });
        
        return interaction.editReply({
          content: 'Error: Inner diameter must be less than outer diameter.'
        });
      }
      
      // Create parameters object
      const params = {
        outer_diameter: outerDiameter,
        inner_diameter: innerDiameter,
        thickness: thickness
      };
      
      await interaction.editReply(
        `Generating washer model with:\nOuter Diameter: ${outerDiameter}\nInner Diameter: ${innerDiameter}\nThickness: ${thickness}`
      );
      
      try {
        // Generate the model
        const result = await cadService.generateModel('washer', params);
        logger.info('Model generation successful', { result });
        
        // Get the filename
        const fileName = result.fileName;
        
        // Try to render the model and get an image
        let imagePath = null;
        try {
          logger.info('Attempting to render the model', { fileName });
          imagePath = await cadService.renderModel(fileName, 'washer');
          logger.info('Rendering successful', { imagePath });
          
          // Send the rendered image
          const attachment = new AttachmentBuilder(imagePath, { name: 'washer_render.png' });
          
          await interaction.editReply({
            content: `Washer model generated successfully!\n` +
                    `Parameters: Outer Diameter: ${outerDiameter}, Inner Diameter: ${innerDiameter}, Thickness: ${thickness}`,
            files: [attachment]
          });
          
          logger.info('Sent message with rendered image');
        } catch (renderError) {
          logger.warn('Rendering failed, sending text-only response', { 
            error: renderError.message 
          });
          
          // Just send model info if rendering failed
          await interaction.editReply({
            content: `Washer model generated successfully!\n` +
                    `Parameters: Outer Diameter: ${outerDiameter}, Inner Diameter: ${innerDiameter}, Thickness: ${thickness}\n` +
                    `Model file: ${fileName}\n\n` +
                    `(Failed to render model image: ${renderError.message})`
          });
        }
      } catch (error) {
        throw error;
      }
    } catch (error) {
      await handleCommandError(error, interaction, 'washer', {
        params: {
          outerDiameter: interaction.options.getNumber('outer_diameter'),
          innerDiameter: interaction.options.getNumber('inner_diameter'),
          thickness: interaction.options.getNumber('thickness')
        }
      });
    }
  },
};